SingularityService/src/main/java/com/hubspot/singularity/SingularityLeaderController.java

```diff
@@ -71,6 +81,39 @@ public void start() {
 
   public void stop() {
     statePoller.finish();
+    TIMER.cancel();
+  }
+
+  @Override
+  public void stateChanged(CuratorFramework client, ConnectionState newState) {
+    LOG.info("Received update for new zk connection state {}", newState);
+    stateHandlerLock.lock();
+    try {
+      scheduler.setZkConnectionState(newState);
+      if (!newState.isConnected()) {
+        LOG.info("No longer connected to zk, pausing scheduler actions and waiting up to {}ms for reconnect",
+            configuration.getZooKeeperConfiguration().getAbortAfterConnectionLostForMillis());
+        lostConnectionStateChecker = new TimerTask() {
+          @Override
+          public void run() {
+            stateHandlerLock.lock();
```

Comment:
There's a small race here. If a reconnect happens and acquires the lock, then this TimerTask executes and begins waiting for the lock. The cancel does not interrupt the TimerTask, so it will still execute the abort. You'd probably need to some sort of boolean guard to this, because TimerTask doesn't expose whether or not its been cancelled.

Reply:
Added a re-check of the current connection state after acquiring the lock, that should prevent this